### ISSUE TITLE
Request: Support Python's requirements.txt

### ISSUE THREAD

**Issue by presidentbeef** [dependency-check/DependencyCheck#548]
`requirements.txt` is typically located in the root of an application. The file format is [documented here](https://pip.pypa.io/en/stable/reference/pip_install/#requirements-file-format).

Examples:
- [scrapy](https://github.com/scrapy/scrapy/blob/master/requirements.txt)
- [boto](https://github.com/boto/boto/blob/develop/requirements.txt)
- [PokemonGo-Bot](https://github.com/PokemonGoF/PokemonGo-Bot/blob/master/requirements.txt)
- [requests](https://github.com/kennethreitz/requests/blob/master/requirements.txt)

`requirements.txt` may point to other requirements files like [this example](https://github.com/fabric/fabric/blob/master/requirements.txt).

**Comment by jeremylong** [comment]
Thanks for the info - I'll be able to work on this after JavaOne.

**Comment by jborrey** [comment]
💯 @presidentbeef to this idea to this idea! Although the `requirements.txt` file does not capture transitive dependencies and so I think it might be better to first look at what [pipdeptree](https://github.com/naiquevin/pipdeptree) is doing to generate all the direct and transitive dependencies and go from there.

Try it out on a python repo:

``` bash
pip install pipdeptree
pipdeptree
pipdeptree --freeze
pipdeptree --json
```

**Comment by presidentbeef** [comment]
I think `pip freeze` and `requirements.txt` are the same format and could be consumed with the same code in DependencyCheck.

**Comment by jborrey** [comment]
Yeah from my little experience in Python dev that seems to be the case. Quickly testing on some repos though shows that `pipdevtree` is finding a few more dependencies.

From a recent test

``` ruby
[13] pry(main)> arr2.uniq - arr1
=> ["setuptools", "rlp", "distribute", "pip", "wheel", "wsgiref"]
```

`arr2` was generated from `pipdevtree --freeze` and `arr1` from `pip freeze`.

**Comment by nowaktz** [comment]
@jeremylong What's the current status of `requirements.txt` support?

**Comment by jeremylong** [comment]
@nowaktz unfortunately - still pending. I've been dealing with other issues. I'm hoping I can get to expanding/improving the python support soon.

**Comment by wkoot** [comment]
As explained in your [comment on issue 963](https://github.com/jeremylong/DependencyCheck/issues/963#issuecomment-342019128), the `__init__.py` files are inspected in particular.
For completeness, the file referred to was renamed - https://github.com/jeremylong/DependencyCheck/blob/master/core/src/main/java/org/owasp/dependencycheck/analyzer/PythonPackageAnalyzer.java

Note that [PEP420](https://www.python.org/dev/peps/pep-0420/) which is implemented since Python 3.3, allows the `__init__.py` files to be dropped completely,
It would be great to see the Python support improve properly, but it seems this throws another monkey wrench into the problem.
Perhaps the PythonDistributionAnalyzer can be extended to comprehend requirements.txt instead?

**Comment by MuhaddiMu** [comment]
Any updates on this?

**Comment by jhermann** [comment]
Just FYI, requirements are slowly phased out towards ``pyproject.toml`` and/or ``Pipfile[.lock]``.

A successful strategy to solve this is to delegate to existing tools instead of (bad / partial) re-implementations. Delegation can be using ``pip freeze`` with little requirements towards tool support (pip is there anyway), or else use more versatile but less common tools like https://github.com/dephell/dephell (which has a stated goal of being able to convert between all the metadata representations in the Python ecosystem).

Letting dephell create a ``Pipfile.lock`` and then parse that and only that would lower efforts, and equally important keep the cat&mouse game out of *this* project and where it belongs (within the Python ecosystem).

`pipdeptree` has a problem, namely it cannot differentiate between environments / dependency sets for a specific purpose (runtime / test / qa / docs / tools / …).

**Comment by jhermann** [comment]
BTW, you could also change the title of this issue to "support Pipfile.lock" and then we could agree I create that file when missing in https://github.com/jhermann/dependency-check-py before spawning the DependencyCheck process.

This has the highest chance of timely delivery of broad Python support, especially when it comes to ‘modern’ projects.

Example (`dephell` applied to itself):

    dephell convert --from=pyproject.toml --to-path=Pipfile --to-format=pipfile \
    && pipenv lock --pre

This creates:

    {
        "_meta": {
            "hash": {
                "sha256": "9ba6ba69ebca0aaf5cc41117d8cb674908ad0c23f543808af3c78e14289a0a11"
            },
            "pipfile-spec": 6,
            "requires": {
                "python_version": "3.6"
            },
            "sources": [
                {
                    "name": "pypi",
                    "url": "https://pypi.org/simple/",
                    "verify_ssl": true
                }
            ]
        },
        "default": {
            "aiofiles": {
                "hashes": [
                    "sha256:021ea0ba314a86027c166ecc4b4c07f2d40fc0f4b3a950d1868a0f2571c2bbee",
                    "sha256:1e644c2573f953664368de28d2aa4c89dfd64550429d0c27c4680ccd3aa4985d"
                ],
                "index": "pypi",
                "version": "==0.4.0"
            },
            …
        },
        "develop": {
            "aiohttp": {
                "hashes": [
                    "sha256:173267050501e1537293df06723bc5e719990889e2820ba3932969983892e960",
                    "sha256:438f1f1555c02c50894604d94944cff188fe138b46467b7fa99fdceb51ab5842",
                    "sha256:90bed250d1435aef33a1f8c439c5056d5d25a44fe6caf33fcafafed805bad4dc",
                    "sha256:93c3b14747413f38f094a60e98f55e73831f0c9a23ae7faa3dc97d8963e13021",
                    "sha256:a6e70a38d883185b1921d8122759661c39ade54949770394412a9e713fec6fa7",
                    "sha256:b5036133c1ba77ed5a70208d2a021a90b76fdf8bf523ae33dae46d4f4380d86f",
                    "sha256:c138451a82cdbf65cddf952941d5c7a1a2cac8ce3bc618dee8d889e5251ec7a5",
                    "sha256:c94770383e49f9cc5912b926364ad022a6c8a5dbf5498933ca3a5713c6daf738",
                    "sha256:ea26536ae06df6dac021303a0df72c79e55512070e6a304ba93ad468a3a754dc"
                ],
                "index": "pypi",
                "version": "==4.0.0a1"
            },
            …
            "zipp": {
                "hashes": [
                    "sha256:ccc94ed0909b58ffe34430ea5451f07bc0c76467d7081619a454bf5c98b89e28",
                    "sha256:feae2f18633c32fc71f2de629bfb3bd3c9325cd4419642b1f1da42ee488d9b98"
                ],
                "version": "==2.1.0"
            }
        }
    }

So then all you need is a JSON parser and some logic to work on the data.

**Comment by fcano** [comment]
In order to support Pipfile or Pipfile.lock, what needs to be done will be very similar to #2526 , right? It would be needed to change the name of the analyzer, the regex or function that detects the versions and the name of the files to look for.
